File: 26_string_manipulation/26_2_string_join.py
def join(arr: list, s: str) -> str:
  if len(arr) == 0:
    return ""
  if len(arr) == 1:
    return arr[0]

  joined_arr = []

  for i in range(len(arr)):
    joined_arr.append(arr[i])
    if i < len(arr)-1:
      joined_arr.append(s)

  return "".join(joined_arr)  

if __name__ == "__main__":
  # Example1
  arr = ["join", "by", "space"] 
  s = " "
  print(join(arr, s))

  # Example2
  arr = ["b", "", "k", "", "p", "r n", "", "d", "d!!"]
  s = "ee"
  print(join(arr, s))

File: 26_string_manipulation/26_2_string_join.cpp
#include <iostream>
#include <vector>
#include <string>

std::string join(std::vector<std::string> arr, std::string s) {
  if (arr.size() == 0) {
    return "";
  }

  if (arr.size() == 1) {
    return arr[0];
  }

  std::string result = "";

  for (int i = 0; i < arr.size(); i++) {
    result += arr[i];
    if (i < arr.size() - 1) {
      result += s;
    }
  }

  return result;
}

int main() {
  // Example1
  std::vector<std::string> arr1 = {"join", "by", "space"};
  std::string s1 = " ";
  std::cout << join(arr1, s1) << std::endl;

  // Example2
  std::vector<std::string> arr2 = {"b", "", "k", "", "p", "r n", "", "d", "d!!"};
  std::string s2 = "ee";
  std::cout << join(arr2, s2) << std::endl;

}